File: lib/languages/python/type-annotations.py
# Basic type annotations (Python 3.5+)
name: str = "Alice"
age: int = 30
height: float = 1.75
is_active: bool = True

# Function type annotations
def greet(name: str) -> str:
   return f"Hello, {name}"

# Optional types
from typing import Optional
def find_user(user_id: int) -> Optional[dict]:
   if user_id == 0:
       return None
   return {"id": user_id, "name": "User"}

# Union types
from typing import Union
def process(value: Union[int, str]) -> str:
   return str(value)

# Type aliases
from typing import List, Dict
UserId = int
UserName = str
Users = Dict[UserId, UserName]

# Generic types
from typing import TypeVar, Generic, List
T = TypeVar('T')

class Stack(Generic[T]):
   def __init__(self) -> None:
       self.items: List[T] = []
   
   def push(self, item: T) -> None:
       self.items.append(item)
   
   def pop(self) -> T:
       return self.items.pop()

# Using the generic class
int_stack: Stack[int] = Stack()
str_stack: Stack[str] = Stack()

File: lib/languages/kotlin/type-annotations.kt
// Basic type annotations
val name: String = "Alice"
val age: Int = 30
val height: Double = 1.75
val isActive: Boolean = true

// Function type annotations
fun greet(name: String): String {
   return "Hello, $name!"
}

// Type aliases
typealias UserId = Int
typealias UserName = String

// Generic type aliases
typealias StringList = List<String>
typealias Predicate<T> = (T) -> Boolean

// Function types
val onClick: () -> Unit = { println("Clicked!") }
val transform: (Int) -> String = { it.toString() }
val biFunction: (String, Int) -> Boolean = { s, i -> s.length > i }

// Nullable types
val nullableString: String? = null
val nonNullString: String = "Cannot be null"

// Collection types
val numbers: List<Int> = listOf(1, 2, 3)
val mutableNumbers: MutableList<Int> = mutableListOf(1, 2, 3)
val nameToAge: Map<String, Int> = mapOf("Alice" to 30, "Bob" to 25)

// Array types
val intArray: IntArray = intArrayOf(1, 2, 3)
val stringArray: Array<String> = arrayOf("a", "b", "c")

// Type projections
fun copy(from: Array<out Any>, to: Array<Any>) {
   // 'out' makes 'from' covariant (read-only)
}

// Star projection
fun printAll(list: List<*>) {
   // '*' is like '?' in Java, represents an unknown type
   list.forEach { println(it) }
}

